### DISCUSSION TITLE
Application level data reception acknowledgements

### DISCUSSION THREAD

**Discussion by valerijbielskij** [OpenDDS/OpenDDS#3568] [Q&A] [answered]
Hello, I'm trying to build a distributed message queue with point to point messaging using OpenDDS. In order to achieve it I have the following reader/writer qos setup:
`qos.durability.kind = DDS::DurabilityQosPolicyKind::TRANSIENT_LOCAL_DURABILITY_QOS;`
`qos.reliability.kind = DDS::ReliabilityQosPolicyKind::RELIABLE_RELIABILITY_QOS;`
`qos.history.kind = DDS::HistoryQosPolicyKind::KEEP_ALL_HISTORY_QOS;`

One to one communication is enforced either by utilizing ContentFilteredTopic or by creating unique topic names for each communication pair. I'm planning to assign distinct key for each data sample so that it could be disposed using same key once reader acknowledges reception (key will be also used to retrieve sample from the writer in case sample has to be put into a different queue).

My problem is that I'm unable to find meaningful way to achieve application level acknowledgements with OpenDDS, for example RTI seemly has exactly what I'm looking for [here](https://community.rti.com/static/documentation/connext-dds/5.2.3/doc/manuals/connext_dds/html_files/RTI_ConnextDDS_CoreLibraries_UsersManual/Content/UsersManual/Application_Acknowledgment.htm). I know that I could just implement explicit ACK messages as data samples but this is a bit cumbersome and probably way less efficient.

How difficult would it be to extend OpenDDS with RTI like callback function for application acknowledgements? Any hints or guidelines are very appreciated.

**Comment by jrw972** [comment, marked as answer]
Extending OpenDDS with this kind of functionality will be difficult because it's going to require touching a lot of different pieces.

First, you will need to extend the IDL with the necessary QoS policies.  The QoS policies will need to be passed down to discovery, communicated to the peer, and the matching logic updated.  You will need to address the probably interoperability, e.g., what happens when one side doesn't understand the QoS.

Second, you will need to add the DataReader methods and update the transports to support the QoS policies.  This will involve changing the implementation of reliability in the Rtps transport so that the reader does not automatically acknowledge a sample but waits for explicit acknowledgement.

Third, you will need to implement the callback on the DataWriter side.

Fourth, you will need to consider all of the corner cases laid out by RTI.

**Reply by mitza-oci** [reply]
In case it wasn't already clear (or may not be clear to others reading this), the "application acknowledgement" feature under discussion is not part of the DDS standard.  That doesn't mean it's not useful, or that it couldn't ever be part of OpenDDS, but it may provide context on why it's not already provided.

**Reply by valerijbielskij** [reply]
> Extending OpenDDS with this kind of functionality will be difficult because it's going to require touching a lot of different pieces.
> 
> First, you will need to extend the IDL with the necessary QoS policies. The QoS policies will need to be passed down to discovery, communicated to the peer, and the matching logic updated. You will need to address the probably interoperability, e.g., what happens when one side doesn't understand the QoS.
> 
> Second, you will need to add the DataReader methods and update the transports to support the QoS policies. This will involve changing the implementation of reliability in the Rtps transport so that the reader does not automatically acknowledge a sample but waits for explicit acknowledgement.
> 
> Third, you will need to implement the callback on the DataWriter side.
> 
> Fourth, you will need to consider all of the corner cases laid out by RTI.

Thanks for the quick reply. I think that point 1 is optional since our QoS is sort of static and is not expected to change during runtime. As for the point 2 it's probably also acceptable that reader would automatically acknowledge even before application retrieves the sample.

Point 3 is what I'm actually interested in. I assume that it has to do with dds\DCPS\DataWriterImpl.h/.cpp and maybe also dds\DCPS\WriterDataContainer.h/.cpp. What I've tried so far is to debug WriteDataContainer::data_delivered, however it get's triggered on every write even if there is no associated data reader which I find a bit confusing. Does it happen because there is no active connections hence the implicit acknowledgement?

**Comment by mitza-oci** [comment]
Please add some more detail around how the high-level description "distributed message queue" maps to topics/readers/writers/QoS, etc.  We can try to make some suggestions based on that.

**Reply by valerijbielskij** [reply]
Imagine we have a constellation of 3 nodes, A, B and C. A needs to transfer data to C but it's never going to have such possibility to do it directly, it will however at some point have contact with B which in turn will later have contact with C. We have a generic data type which consist of packet header and payload (sequence of octets). Since we know that A will at some establish connection with B we want to immediately put data into a topic say "A_to_B" where it get's queued until the actual connection happens, once B retrieves the data it will put into topic "B_to_C". Main QoS was already mentioned in the original question and of course I'm simplifying it a bit but I believe it should be enough to understand the idea. I'm also completely fine to hear that DDS is not meant to solve such problems :)

**Reply by mitza-oci** [reply]
OK, as you've already noted, the Durability QoS Policy helps to solve this problem.  The durable writer in A stores some data in its cache that's later transferred to B.  And the same thing can happen for B -> C.  Since you're asking about application-level acknowledgement, it seems like the missing piece is when to tell the durable writers that they can remove data from their caches.  Otherwise that data will just accumulate indefinitely.

Here are a few mechanisms to consider:
* There are QoS policies that are designed to bound the cache in the writer either based on space or on time (History, Resource Limits, Lifespan)
* Discovery-based events like "publication matched" / "subscription matched" are made available to applications
* Discovery can also be used to transmit application-specific metadata via the User Data QoS Policy (the values appear in the standard Built-In Topics)

If you're letting the QoS policies manage the cache for you, there may not much more to do.  If not, will the reader have enough information to know when to disassociate?  That event could help the writer determine when to explicitly dispose an instance.  Changing the User Data gives you a built-in channel to get feedback from B to A.  And you've already mentioned the "explicit ACK messages" which may be a good last resort -- it's similar to the pattern employed by the RPC-over-DDS spec.

**Reply by valerijbielskij** [reply]
Thanks for suggesting User Data QoS Policy, I haven't consider using that! As for the immediate disposal after disassociation - I think such approach is not bullet proof enough, there might be connection losses. After reconnect reader doesn't want to receive samples which it already processed during previous connection, similarly, writer should not dispose all data just because of disassociation, it needs to know exactly how much data was received so that pending data could be requeued again (possibly with a different contact, through a different topic).

While typing this I realized that since duration of contact window is known maybe I could utilize TCP/IP configuration options like `active_conn_timeout_period` and `passive_reconnect_duration` to avoid situations where reader after reconnection receives samples which it already processed? This would allow to skip sample disposal from the writer side, I reckon disposal is an expensive operation.

**Reply by valerijbielskij** [reply]
> 

Apparently my assumption about transport configuration was wrong, from what I have observed aforementioned parameters do not influence data redistribution, you data reader will still receive all of the samples again even if the timeout hasn't not been reached yet.

**Comment by valerijbielskij** [comment]
This is what I ended up with:
1. Extended DDS::DataWriterListener with 
    `void on_reception_acknowledged(::DDS::DataWriter_ptr writer, ::DDS::InstanceHandle_t handle);`
2. In `DataWriterImpl::data_delivered(const DataSampleElement* sample)` added following code:
```
auto listener = listener_for(DEFAULT_STATUS_MASK);
 if (listener)
  {
    if (sample->get_header().message_id_ == SAMPLE_DATA && sample->get_handle())
    {
        listener->on_reception_acknowledged(this, sample->get_handle()->instance_handle_);
    }
}
```
Now this seems to work but I haven't tested it thoroughly yet.
Note that `data_delivered` is invoked by the transport even when there is no actually connected data readers, in such case `TransportClient::send_i` can be modified to either skip such cases or invoke  `data_delivered` with some additional flag.
